// fpdb/index.js

const fs = require("fs")
var path = require("path")
var v = "1.0.2"



//text encoder
function enc(data){
return Buffer.from(data, "utf-8").toString("base64")
}
function dec(data){
return Buffer.from(data, "base64").toString("utf-8")
}

//db file creator and checker
try {
  if(fs.existsSync("./fpdb/db/")){
    console.log("\x1b[32m Database Directory Check Complete \x1b[0m")
  } else {
    initialize()
    console.log("\x1b[32m Database Directory Check Complete \x1b[0m")
  }
} catch (error) {
  console.log(`\x1b[31m Error In FPDB While Running Database Directory Check : \n ${error} \x1b[0m`)
}
try {
  fs.readFileSync("./package.json")
} catch (error) {
  console.log("y")
}


// version checker
const http = require("https");
function vcheck(){
const options = {
  "method": "POST",
  "hostname": "chatroom.dhruvakalur.repl.co",
  "port": null,
  "path": "/fpdb/",
  "headers": {
    "Accept": "*/*",
    "User-Agent": "Thunder Client (https://www.thunderclient.com)"
  }
};

const req = http.request(options, function (res) {
  const chunks = [];

  res.on("data", function (chunk) {
    chunks.push(chunk);
  });

  res.on("end", function () {
    const body = Buffer.concat(chunks);
    if(body.toString()!=v){
        console.log("\u001b[33m Warning: Your FPDB Version Is Outdated, Please Install The New Version via npm i fpdb@latest \u001b[0m")
    } else{
      console.log(`\x1b[32m Version Is Latest \x1b[0m`)
    };
  });
});

req.end()
}
function checkInternet(cb) {
  console.log("\x1b[32m Checking If The Package(FPDB) Has The Latest Version \x1b[0m")
  require('dns').lookup('google.com',function(err) {
      if (err && err.code == "ENOTFOUND") {
          cb(false);
      } else {
          cb(true);
      }
  })
}

checkInternet(function(isConnected) {
  if (isConnected) {
      vcheck()
  } else {
      console.log(" \u001b[33m Warning: Couldn't Check For Latest Version Of FPDB Due To Lack Of \n Internet, Nothing To Worry. \u001b[0m")
  }
});


// functions 

//initiation function
exports.init = function(){
  if(initialize()==undefined){
    console.log("Initiation Done")
  } else {
    console.log("Initiation Not Complete")
  }
}

// Set() Function
exports.set=function(key,cname,value){

    key = ""+key+"";value=""+value+""
    value = enc(value), key = enc(key), cname = enc(cname);
        if(cname){
          if(fs.existsSync("./fpdb/db/"+cname+"/")==true){
    fs.writeFile("./fpdb/db/"+cname+"/"+key+'.txt', value, function(err) {
        if (err) {
         initialize()
         console.log("\x1b[31m Couldn't Create Key: "  + dec(key) + ", Please Run Initiation Function \n If This Is Your First Time Using This Package, This Is Normal But If Problem Persists Please Send An E-Mail To dhruvakalur9@gmail.com \x1b[0m")
        }})
      } else {
        console.log(`\x1b[31m The Collection: "${dec(cname)}" Doesn't Exist \x1b[0m`)
      }
      }else{
        fs.writeFile("./fpdb/db/"+key+'.txt', value, function(err) {
          if (err) {
           initialize()
           console.log("\x1b[31m Couldn't Create Key: "  + dec(key) + ", Please Run Initiation Function \n If This Is Your First Time Using This Package, This Is Normal But If Problem Persists Please Send An E-Mail To dhruvakalur9@gmail.com \x1b[0m")
          }}) 
      }
    }


//CreateCollection() Function
exports.createCol = function(cname){
  cname = ""+cname+"";
  cname = enc(cname)
  try {
    fs.mkdirSync('./fpdb/db/'+cname+"/")
  } catch (error) {
    if(`${error}`.startsWith("Error: EEXIST: ")){
      console.log("\x1b[31m The Collection '"+cname+"' Is Already Created \x1b[0m")
        } else{
   console.log("\x1b[31m Error While Creating Collection '"+dec(cname)+"', Error : \n"+ error)
    }
  }
 

} 




    // helper for initiation function
    function initialize(){

    fs.mkdir('./fpdb/',function(err){
        if(err){
        if (err.message.startsWith("EEXIST:") != true) {
            
        return `${console.log("Error Occurred In Initiation : \n\n" +err)}`
        }
        fs.mkdir('./fpdb/db/',function(err){
          if(err){
          if (err.message.startsWith("EEXIST:") != true) {
              
          return `${console.log("Error Occurred In Initiation : \n\n" +err)}`
          }
        }})
      }
    
    }, )
}



// get() function
exports.get = (key,cname,callback)=>{
  setTimeout(()=>{
    if(cname){
      if(fs.existsSync('./fpdb/db/'+enc(cname)+"/")==true){
    fs.readFile('./fpdb/db/'+enc(cname)+"/"+enc(key)+".txt", (err, data) => {
      if (err){
        if(`${err}`.startsWith("Error: ENOENT: ")){callback(` Error: The Key : "${key}" Doesn't Exist`,undefined)}else{callback("\x1b[31m Error Occurred While Accessing Key :"+key,undefined)}
      } else {
      callback(undefined,dec(data.toLocaleString()));
      }
    });
  }else{
    console.log(`\x1b[31m The Collection: "${cname}" Doesn't Exist \x1b[0m`)
  }
  }else{
    fs.readFile('./fpdb/db/'+enc(key)+".txt", (err, data) => {
      if (err){
        if(`${err}`.startsWith("Error: ENOENT: ")){callback(` Error: The Key : "${key}" Doesn't Exist`,undefined)}else{callback("\x1b[31m Error Occurred While Accessing Key :'"+key+"', Error:\n"+err)}
      } else {
      callback(undefined,dec(data.toLocaleString()));
      }
    });
  }
  
},10)
}

//deleteall() function
exports.delAll=()=>{
  if(fs.existsSync("./fpdb/db/")){
  fs.rm('./fpdb/db/', { recursive:true }, (err) => {
    if(err){
        // File deletion failed
        console.error(err.message);
        return;
    }
    initialize();
})
} else {
  console.log(`\x1b[31m Error: You Have Not Run The Initialisation Function, Visit The README.MD(Docs) Here https://www.npmjs.com/package/fpdb`)
}
}


//DeleteKey() Function
exports.delKey=async function(key,cname){
  if(cname){
    if(fs.existsSync('./fpdb/db/'+enc(cname)+"/"+enc(key)+".txt")) {
      try {
        await fs.unlink('./fpdb/db/'+enc(cname)+"/"+enc(key)+".txt");
      } catch (error) {
        console.error('\x1b[31m Error Occurred In FPDB While Deleting Key: '+key+' \n', error.message + "\x1b[0m");
      }
      } else {
        console.log(`\x1b[31m Error: This Key Does Not Exist, Visit The README.MD(Docs) Here https://www.npmjs.com/package/fpdb`)
      }
  }else{

  
if(fs.existsSync('./fpdb/db/'+enc(key)+".txt")) {
try {
  await fs.unlink('./fpdb/db/'+enc(key)+".txt");
} catch (error) {
  console.error('\x1b[31m Error Occurred In FPDB While Deleting Key: '+key+' \n', error.message + "\x1b[0m");
}
} else {
  console.log(`\x1b[31m Error: This Key Does Not Exist, Visit The README.MD(Docs) Here https://www.npmjs.com/package/fpdb`)
}
}
}

//deletecollection() Function
exports.delCol = (cname)=>{
if(fs.existsSync('./fpdb/db/'+enc(cname)+"/")){
  fs.rm('./fpdb/db/'+enc(cname)+"/", { recursive:true }, (err) => {
    if(err){
        // File deletion failed
        console.error(err.message);
    }
})
} else {
  console.log(`\x1b[31m Error: This Collection Does Not Exist, Visit The README.MD(Docs) Here https://www.npmjs.com/package/fpdb`)
}
}

//listkeys() function
exports.listKeys=(cname,callback)=>{
  if(cname){
    fs.readdir("./fpdb/db/"+enc(cname)+"/", (err, files) => {
      if (err) {
        if(`${err}`.startsWith("Error: ENOENT: ")){
          console.log("\x1b[31m The Collection '"+cname+"' Does Not Exist \x1b[0m")
        }else{
          callback(err,undefined);
        }
      } else {
        files.forEach(file => {
          if (path.extname(file) == ".txt")
            callback(undefined,dec(file.replace(".txt","")))
        })
      }
    })
  } else {
  fs.readdir("./fpdb/db/", (err, files) => {
    if (err) {
      if(`${err}`.startsWith("Error: ENOENT: ")){
        console.log("\x1b[31m The Collection '"+cname+"'Does Not Exist \x1b[0m")
      }else{
        callback(err,undefined);
      }
    } else {
      files.forEach(file => {
        if (path.extname(file) == ".txt")
          callback(undefined,dec(file.replace(".txt","")))
      })
    }
  })
}
}

//listcollections helper
function getDirectories(path) {
  return fs.readdirSync(path).filter(function (file) {
    return fs.statSync(path+'/'+file).isDirectory();
  });
}

//listcollections() function
exports.listCol=(callback)=>{
  if(fs.existsSync("./fpdb/db/")){
 try {
  callback(undefined,`${dec(getDirectories("./fpdb/db/").toLocaleString())}`)
 } catch (error) {
  callback(error,undefined)
 }
} else {
  console.log("\x1b[31m Please Run The init() Function, For Docs Click Here : https://www.npmjs.com/package/fpdb \x1b[0m")
}
}
